dhis2/dhis2-core
# feat: job progress tracking for sync jobs [DHIS2-12858]

### Summary
This PR adds `JobProgress` tracking to jobs that synchronize data and metadata.

* `MetadataSyncJob`
* `DataSynchronizationJob`
* `EventProgramsDataSynchronizationJob`
* `TrackerProgramsDataSynchronizationJob`

The synchronisation itself is

* `DataValueSynchronization`
* `CompleteDataSetRegistrationSynchronization`
* `EventSynchronization`
* `TrackerSynchronization`

Previously these would use state variable in the `@Component` which is bad practice with regards to muti-threading.
These variables have been turned into immutable context classes instead that are passed as part of method parameters making this safe to use in multi-threaded scenario.

### Automatic Testing
Existing tests were adjusted.

### Manual Testing
* setup a remote server using the system settings app (I abused play/dev)
* open scheduler app, add new job configuration for the synchronisation job to test
* run the added job manually
* check http://localhost:8080/api/scheduling/running and http://localhost:8080/api/scheduling/completed

Such test should be done for 
* data (value) sync
* event sync
* tracker sync
* metdata sync

At the moment these do fail at some point for reasons unrelated to the changes in this PR. For example because not all synchronized values were successfully pushed or because of issues in the service code running queries needed to perform the task.
